node-grove/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+        // concurrently.
+        thread::spawn(move || {
+            let path = Path::new(&path_string);
+            // Open a connection to groveDb, this will be moved to a separate thread
+            // TODO: think how to pass this error to JS
+            let mut grove_db = GroveDb::open(path).unwrap();
+
+            // Blocks until a callback is available
+            // When the instance of `Database` is dropped, the channel will be closed
+            // and `rx.recv()` will return an `Err`, ending the loop and terminating
+            // the thread.
+            while let Ok(message) = rx.recv() {
+                match message {
+                    DbMessage::Callback(callback) => {
+                        // The connection and channel are owned by the thread, but _lent_ to
+                        // the callback. The callback has exclusive access to the connection
+                        // for the duration of the callback.
+                        callback(&mut grove_db, &channel);
+                    }
+                    // Immediately close the connection, even if there are pending messages
+                    DbMessage::Close(callback) => {
+                        callback(&channel);
+                        break;
+                    }
+                }
+            }
+        });
+
+        Ok(Self { tx })
+    }
+
+    // Idiomatic rust would take an owned `self` to prevent use after close
+    // However, it's not possible to prevent JavaScript from continuing to hold a closed database
+    fn close(
+        &self,
+        callback: impl FnOnce(&Channel) + Send + 'static,
+    ) -> Result<(), mpsc::SendError<DbMessage>> {
+        self.tx.send(DbMessage::Close(Box::new(callback)))
+    }
+
+    fn send_to_db_thread(
+        &self,
+        callback: impl FnOnce(&mut GroveDb, &Channel) + Send + 'static,
+    ) -> Result<(), mpsc::SendError<DbMessage>> {
+        self.tx.send(DbMessage::Callback(Box::new(callback)))
+    }
+}
+
+// Ensures that GroveDbWrapper is properly disposed when the corresponding JS object
+// gets garbage collected
+impl Finalize for GroveDbWrapper {}
+
+// External wrapper logic
+impl GroveDbWrapper {
+    // Create a new instance of `Database` and place it inside a `JsBox`
+    // JavaScript can hold a reference to a `JsBox`, but the contents are opaque
+    fn js_open(mut cx: FunctionContext) -> JsResult<JsBox<GroveDbWrapper>> {
+        let grove_db_wrapper =
+            GroveDbWrapper::new(&mut cx).or_else(|err| cx.throw_error(err.to_string()))?;
+
+        Ok(cx.boxed(grove_db_wrapper))
+    }
+
+    fn js_get(mut cx: FunctionContext) -> JsResult<JsUndefined> {
+        let js_path = cx.argument::<JsArray>(0)?;
+        let js_key = cx.argument::<JsBuffer>(1)?;
+        let js_callback = cx.argument::<JsFunction>(2)?.root(&mut cx);
+
+        let path = converter::js_array_of_buffers_to_vec(js_path, &mut cx)?;
```

Comment:
perhaps we need to represent paths as a continuous memory location sometime, otherwise we need to do such acrobatics

Reply:
@fominok It depends, I think from the Rust perspective it looks alright, otherwise you'll have to parse continuous memory to get path fragments for it